Patches from upstream

akleja opened the pull request:
<!--

IMPORTANT! READ

Any spam PRs will be closed.
Please make sure your PR
complies with the requirements.
-->
Mainly commits related to track affordances and labels.
Merge conflicts in TrackPanel due to not including commits to move drawing of edit cursor to TrackArtist.cpp

Resolves: https://github.com/tenacityteam/tenacity/issues/356 https://github.com/tenacityteam/tenacity/issues/354


<!-- Use "x" to fill the checkboxes below like [x]
an asterisk (*) after the entry indicates required
-->

<details>
<summary>Checklist</summary>

- [x] I have signed off my commits using `-s` or `Signed-off-by`\* (See: [Contributing § DCO](https://github.com/tenacityteam/tenacity/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))
- [x] I made sure the code compiles on my machine
- [ ] I made sure there are no unnecessary changes in the code\*
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving\*
- [ ] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"\*

\* indicates required

</details>

Be-ing commented:
Can you clarify how you picked commits? Did you go through the Audacity Git history and only pick the ones you were interested in, or did you pick everything one by one?

akleja commented:
I picked everything related to tracks (affordances, sizing and labels mostly) going back to the last merge. The reason for this is that the current implementation of the affordances has a lot of bugs/unfinished functionality.
I did not include ones where we've diverted, namely commits related to the edit cursor. The only one not directly related is 612cf33 which is related to the lyrics window.
I did not pick commits unrelated to handling of tracks, except the one named above, since I'm not sure which we'd want to merge or not.

panos commented:
For future reference, I tend to maintain the branches `upstream-audacity` and `patches-from-upstream` to simplify the process.

Be-ing commented:
How does that simplify the process?

akleja commented:
@panos Okay, I didn't know. However since I don't have read/write privileges to the repos that workflow doesn't seem simpler to me. Opening pull requests for 3 branches seems like it would complicate things. 
Should I open a pull request for the patches-from-upstream branch as well? Seems otherwise like it only would need to be synced with master if this PR is merged.
I'm not sure about the best way to update the upstream-audacity branch.

Perhaps writing some guidelines for merging upstream patches would be a good thing? But having a workflow that requires r/w privileges or opening 3 PR's might not be the best idea if we want to share the work of keeping up with upstream.
Or perhaps I've misunderstood something about this process?

panos commented:
> How does that simplify the process?

Everything is in a separate branch already and I can gradually move towards the goal of selecting patches that work on top of master before applying them to master itself.

panos commented:
I'm not sure what to do here.

akleja commented:
@panos I don't understand.
Do you mean that there's some problem with the selection of patches or how they're merged?
Or is it that I haven't updated the other branches that's the problem?
If so I'd appreciate some instructions as to a preferred workflow. I don't know how to update the upstream branch in a good way. Also It doesn't make sense to me to update the patches-from-upstream branch before the PR is approved, in case it isn't for some reason.
Or have I've messed up in some other way ?

panos commented:
Oh, sorry for not being clear. Is this ready and good to go (to be merged on top of master), or are there still changes that need to be merged?

panos commented:
Hi, excuse me for the late review, there seems to be a conflict here.

leio commented:
It would be good to have the commit IDs from audacity where it came from in the commit logs, especially if it isn't fully brought up to date (all commits considered since the last full consideration and anything missing are not included on purpose), so that other merge work would know that something is for sure already included on our end when looking why something is conflicting now.
I would do this by using the `-x` option of cherry-pick with audacity repository as a remote in my tenacity local repository, and perhaps editing the standard message it gives (`(cherry picked from commit <full hash>)`) to say `audacity commit` instead of just `commit`.

panos commented:
Just tried rebasing and fixing this, there's some problem with `toolbars/MixerToolBar.cpp` and `toolbars/MixerToolBar.h` that I haven't figured out yet.

panos commented:
Oh right, we removed that with @akleja here: https://github.com/tenacityteam/tenacity/commit/056de91cf79179f53f56be2f809709137d65e2a5

panos commented on src/TrackPanel.cpp:
```diff
@@ -91,6 +93,9 @@ is time to refresh some aspect of the screen.
 #include <wx/dcclient.h>
 #include <wx/graphics.h>
 
+static_assert( kVerticalPadding == kTopMargin + kBottomMargin );
```
This (`a95e4dd31cb770cd1eab87e18b36e035506882ab`) fails.

akleja replied:
That's because we've removed the black border around tracks. The issue is mitigated by setting `kVerticalPadding = 6` in ViewInfo.h which is included in this PR in an additional commit: https://github.com/tenacityteam/tenacity/pull/515/commits/ddbd9ed0cd442cc85984a5d604be4a7b00449bf1

panos reviewed (changes requested):
Tried to apply this because this seemed like a blocker, came across an issue.

panos commented:
Added this pull request as a release blocker, as the current state of our affordances is half-broken/half-finished.

akleja commented:
So at last I've rebased and also added the commit hashes. I did this manually for now since I didn't want to go through resolving merging conflicts again. I hope the formatting in the comments is ok.

panos commented:
I didn't notice this, reviewing ASAP.

akleja commented:
Thank you!

panos reviewed (approved):
Works. Tried to check whether this overwrote any of our own changes, didn't find anything.

panos commented:
Thanks for coming back to help, I'm also undertaking some heavy backporting efforts.

akleja commented:
Thanks, I've had some IRL issues preventing me from participating for a while, but I'll try to gradually increase my efforts again.
Yeah I've seen the beginning of keeping up with the GUI separation, really great initiative!
